compiler/nilcheck.nim

```diff
@@ -0,0 +1,624 @@
+#
+#
+#           The Nim Compiler
+#        (c) Copyright 2017 Andreas Rumpf
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+import 
+  ast, astalgo, hashes, idents, intsets, lineinfos, msgs, options, 
+  renderer, ropes, sequtils, strformat, strutils, tables, treetab, types
+
+# 
+# Env: int => nilability
+# a = b
+#   nilability a <- nilability b
+# deref a
+#   if Nil error is nil
+#   if MaybeNil error might be nil, hint add if isNil
+#   if Safe fine
+# fun(arg: A)
+#   nilability arg <- for ref MaybeNil, for not nil or others Safe
+# map is env?
+# a or b
+#   each one forks a different env
+#   result = union(envL, envR)
+# a and b
+#   b forks a's env
+# if a: code
+#   result = union(previousEnv after not a, env after code)
+# if a: b else: c
+#   result = union(env after b, env after c)
+# result = b
+#   nilability result <- nilability b, if return type is not nil and result not safe, error
+# return b
+#   as result = b
+# try: a catch: b finally: c
+#   in b and c env is union of all possible try first n lines, after union of a and b and c
+# case a: of b: c
+#   similar to if
+# call(arg)
+#   if it returns ref, assume it's MaybeNil: hint that one can add not nil to the return type
+# call(var arg) # zahary comment
+#   if arg is ref, assume it's MaybeNil after call
+# loop
+#   union of env for 0, 1, 2 iterations as Herb Sutter's paper
+# return
+#   if something: stop (break return etc)
+#   is equivalent to if something: .. else: remain
+# new(ref)
+#   ref becomes Safe
+# each check returns its nilability and map
```

Comment:
Why so much commented stuff :grey_question: Kinda bloats the code and is not really doing anything.
Maybe make it into Documentation comments with a `**Note:** Internal compiler documentation.`, or remove it.
:thinking:

Reply:
yes, this is just the old code, now we have a spec, i should remove it indeed!